Ported watabou.noosa.audio to kotlin

Zrp200 opened the pull request.

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+    private val trackQueue = ArrayList<String?>()
+    @JvmField var shuffle = false
+    @Synchronized
+    fun play(assetName: String?, looping: Boolean) {
```
assetName can't be null

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+    @JvmField var trackList: Array<String>? = null
+    lateinit var trackChances: FloatArray
+
+    private val trackQueue = ArrayList<String?>()
```
Strings inside trackQueue can't be null

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+import com.badlogic.gdx.audio.Music
+import com.watabou.noosa.Game
+import com.watabou.utils.Random
+import java.util.*
```
don't use wildcard imports

Zrp200 replied:
fixed

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+    private var looping = false
+
+    var volume = 1f
+        @JvmName("volume") get
```
redundant getter

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+            player?.volume = value
+        }
+
+    @JvmField var trackList: Array<String>? = null
```
can be private

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+        }
+
+    @JvmField var trackList: Array<String>? = null
+    lateinit var trackChances: FloatArray
```
can be private

QuasiStellar replied:
trackList and trackChances can be one map

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+
+    @Synchronized
+    private fun playNextTrack(music: Music) {
+        if (trackList?.isEmpty() != false || music !== player || player!!.isLooping) {
```
!= false?

Zrp200 replied:
if it is null or true.

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+            }
+            if (shuffle) trackQueue.shuffle()
+        }
+        if (!isEnabled || trackQueue.isEmpty()) {
```
already checked for isEmpty

Zrp200 replied:
this was present in the original code. I'll fix it now though.

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+
+    @Synchronized
+    private fun play(track: String?, listener: Music.OnCompletionListener?) {
+        player = null // this deactives the volume setter logic.
```
why is this needed?

Zrp200 replied:
what is 'this'

QuasiStellar replied:
Deactivating volume setter logic

Zrp200 replied:
done

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+        }
+
+    @get:Synchronized
+    val isPlaying get() = player?.isPlaying == true
```
== true?

Zrp200 replied:
original code returned false if it was null.

QuasiStellar replied:
Ah, right

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+    @Synchronized
+    private fun play(track: String?, listener: Music.OnCompletionListener?) {
+        player = null // this deactives the volume setter logic.
+        kotlin.runCatching {
```
kotlin is redundant

Zrp200 replied:
it's a different method. I didn't need `this` to be bound.

Zrp200 replied:
I'll fix it anyway, but it's redundant.

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+        player = null // this deactives the volume setter logic.
+        kotlin.runCatching {
+            player = Gdx.audio.newMusic(Gdx.files.internal(track)).apply {
+                isLooping = this@MusicPlayer.looping
```
this@MusicPlayer is redundant

Zrp200 replied:
okay, if you say so. Figured it would be nice for consistency.

QuasiStellar replied:
Makes sense, better leave it then

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+import java.util.HashSet
+import kotlin.math.max
+
+object Sample {
```
can be renamed to SamplePlayer

Zrp200 replied:
I'm not in the mood to rename all uses. You can do it if you like.

QuasiStellar replied:
Isn't it automatic?

Zrp200 replied:
maybe but it makes the pull request larger for little reason.

QuasiStellar replied:
Well, that's not a problem

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+            ?: -1
+    }
+
+    private data class DelayedSoundEffect(var id: Any? = null,
```
styleguide:
```
private data class DelayedSoundEffect(
    var id: Any? = null,
    var delay: Float = 0f,
    var leftVol: Float = 0f,
    var rightVol: Float = 0f,
    var pitch: Float = 0f
)

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Synchronized
+    fun pause() { player?.pause() }
```
block-like methods should always have linebreaks

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+            player?.volume = value
+        }
+
+    @JvmField var trackList: Array<String>? = null
```
var annotations should be on their own lines

Zrp200 replied:
okay. I'll put it how you like it.

QuasiStellar replied:
That's not me,  that's standard Kotlin linter

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+    fun play(assetName: String?, looping: Boolean) {
+
+        //iOS cannot play ogg, so we use an mp3 alternative instead
+        var assetName = assetName
```
name shadowed

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+
+    @Synchronized
+    fun playTracks(tracks: Array<String>?, chances: FloatArray, shuffle: Boolean) {
+        if (tracks == null || tracks.isEmpty() || tracks.size != chances.size) {
```
can be replaced with isNullOrEmpty()

Zrp200 replied:
oh, huh, didn't know that one existed.

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+    private fun play(track: String?, listener: Music.OnCompletionListener?) {
+        player = null // this deactives the volume setter logic.
+        kotlin.runCatching {
+            player = Gdx.audio.newMusic(Gdx.files.internal(track)).apply {
```
There is a function for getting internal files: getAsset

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+            override fun run() {
+                for (asset in toLoad) {
+                    val newSound = Gdx.audio.newSound(Gdx.files.internal(asset))
+                    synchronized(this@Sample) { ids[asset] = newSound }
```
getAsset

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+
+object Sample {
+
+    private var ids = HashMap<Any?, Sound>()
```
should be String instead of Any?

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @JvmStatic @Synchronized
+    fun pause() { ids.values.forEach(Sound::pause) }
```
linebreaks

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @JvmStatic @Synchronized
+    fun unload(src: Any?) { ids.remove(src)?.dispose() }
```
src is String

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+    @JvmStatic @JvmOverloads
+    fun play(id: Any?, volume: Float = 1f, pitch: Float = 1f) = play(id, volume, volume, pitch)
+    @JvmStatic @Synchronized
+    fun play(id: Any?, leftVolume: Float, rightVolume: Float, pitch: Float): Long {
```
id is String

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+package com.watabou.noosa.audio
+
+import kotlin.jvm.Synchronized
+import com.watabou.noosa.audio.Sample
```
unused imports

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+            play(id, leftVolume, rightVolume, pitch)
+            return
+        }
+        synchronized(delayedSFX) { delayedSFX.add(DelayedSoundEffect(
```
formatting

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,129 @@
+                                          var pitch: Float = 0f)
+
+    @JvmOverloads @JvmStatic
+    fun playDelayed(id: Any?, delay: Float, volume: Float = 1f, pitch: Float = 1f) = playDelayed(id, delay, volume, volume, pitch)
```
id is String

QuasiStellar reviewed (changes requested):
Rename the commit "remaster: ..." to keep the consistency.
I will make a PR template sometime soon.

Zrp200 commented:
Will fix these later today.

Zrp200 committed:
Adjusted PR based on comments

General:
* removed unneeded imports
* replaced Gdx.internal calls with getAsset
* annotations are now on their own lines
* block methods now are no longer on a single line

MusicPlayer.kt
* removed unused getter for volume
* tracklist is now private, removed JvmField annotation
* play:
   - assetName, track is now not nullable
   - removed name shadowing of assetName
* used `isNullOrEmpty` instead of `?.isEmpty`, fixed some bad logic checks regarding this
* removed an unneeded check for trackqueue being empty in playNextTrack
* changed a kotlin.runCatching call to just being a standard runCatching

Sample.kt:
* ids now has keys of type String, methods changed accordingly.
* reformatted, removed default type arguments for DelayedSoundEffect. Its id is now a String.

Zrp200 commented:
It currently appears to not entirely work, hold on while I look into it.

Zrp200 committed:
merged trackList and trackChances into trackMap

* Added an additional play method that uses the existing track map

Zrp200 committed:
Fixes to Sample.kt

* Converted sound passes by reference to block based invocations; this fixes synchronization issues with the associated calls.
* Removed unneeded @JvmStatic annotations

Also:
* moved delayedSFX call up to mirror the original location of the declaration

Zrp200 commented:
okay it should be fixed now

Zrp200 committed:
Reverted a minor change based on conversations

Zrp200 committed:
refactored play() to remove setting player to null

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/MusicPlayer.kt:
```diff
@@ -0,0 +1,198 @@
+            player?.volume = value
+        }
+
+    private var trackMap: Map<String,Float> = emptyMap()
```
Space after the comma
![image](https://user-images.githubusercontent.com/30569097/185769638-d5a2e748-ee70-43bb-93fc-651ecadd1998.png)

QuasiStellar commented on SPD-classes/src/main/java/com/watabou/noosa/audio/Sample.kt:
```diff
@@ -0,0 +1,131 @@
+
+    @Synchronized
+    fun reset() {
+        for(sound in ids.values) sound.dispose()
```
Space after for, reformat the code before commiting

Zrp200 committed:
Adjusted formatting to be consistent with kotlin style guide

Zrp200 committed:
Ported watabou.noosa.audio to kotlin
